File: benchmark/corpus/javascript/path_safe_1.js
// SAFE: resolved path must stay within BASE_DIR
const express = require('express');
const fs = require('fs');
const path = require('path');
const app = express();

const BASE_DIR = path.resolve('/var/www/files');

app.get('/download', (req, res) => {
    const filename = req.query.file;
    const filepath = path.resolve(path.join(BASE_DIR, filename));
    if (!filepath.startsWith(BASE_DIR + path.sep)) {
        return res.status(403).send('Access denied');
    }
    fs.readFile(filepath, (err, data) => {
        if (err) return res.status(404).send('Not found');
        res.send(data);
    });
});

app.listen(3000);

File: benchmark/corpus/python/path_safe_1.py
"""SAFE: realpath + prefix check prevents path traversal."""
import os
import flask

app = flask.Flask(__name__)
BASE_DIR = os.path.realpath("/var/www/files")


@app.route("/download")
def download():
    filename = flask.request.args.get("file", "")
    filepath = os.path.realpath(os.path.join(BASE_DIR, filename))
    if not filepath.startswith(BASE_DIR + os.sep):
        return flask.abort(403)
    with open(filepath, "rb") as f:
        return flask.Response(f.read(), content_type="application/octet-stream")
